petastorm/spark/spark_dataset_converter.py

```diff
@@ -0,0 +1,289 @@
+    """
+    parsed = urlparse(dir_url)
+    if not parsed.scheme:
+        raise ValueError(
+            'ERROR! A scheme-less directory url ({}) is no longer supported. '
+            'Please prepend "file://" for local filesystem.'.format(dir_url))
+    new_parsed = parsed._replace(path=os.path.abspath(parsed.path))
+    return new_parsed.geturl()
+
+
+def _is_sub_dir_url(dir_url1, dir_url2):
+    """
+    Check whether url1 is a sub directory of url2
+    """
+    url1 = _normalize_dir_url(dir_url1)
+    url2 = _normalize_dir_url(dir_url2)
+
+    parsed1 = urlparse(url1)
+    parsed2 = urlparse(url2)
+
+    return parsed1.scheme == parsed2.scheme and \
+        parsed1.netloc == parsed2.netloc and \
+        parsed1.path.startswith(parsed2.path + os.sep)
+
+
+def _cache_df_or_retrieve_cache_data_url(df, parent_cache_dir_url,
+                                         parquet_row_group_size_bytes,
+                                         compression_codec):
+    """
+    Check whether the df is cached.
+    If so, return the existing cache file path.
+    If not, cache the df into the cache_dir in parquet format and return the
+    cache file path.
+    Use atexit to delete the cache before the python interpreter exits.
+    :param df: A :class:`DataFrame` object.
+    :param parent_cache_dir_url: A string denoting the directory for the saved
+        parquet file.
+    :param parquet_row_group_size_bytes: An int denoting the number of bytes
+        in a parquet row group.
+    :param compression_codec: Specify compression codec.
+    :return: A string denoting the path of the saved parquet file.
+    """
+    # TODO
+    #  Improve the cache list by hash table (Note we need use hash(df_plan +
+    #  row_group_size)
+    with _cache_df_meta_list_lock:
+        df_plan = _get_df_plan(df)
+        for meta in _cache_df_meta_list:
+            if meta.row_group_size == parquet_row_group_size_bytes and \
+                    meta.compression_codec == compression_codec and \
+                    meta.df_plan.sameResult(df_plan) and \
+                    _is_sub_dir_url(meta.cache_dir_url, parent_cache_dir_url):
+                return meta.cache_dir_url
+        # do not find cached dataframe, start materializing.
+        cached_df_meta = CachedDataFrameMeta.create_cached_dataframe(
+            df, parent_cache_dir_url, parquet_row_group_size_bytes,
+            compression_codec)
+        _cache_df_meta_list.append(cached_df_meta)
+        return cached_df_meta.cache_dir_url
+
+
+def _materialize_df(df, parent_cache_dir, parquet_row_group_size_bytes,
+                    compression_codec):
+    uuid_str = str(uuid.uuid4())
+    save_to_dir = os.path.join(parent_cache_dir, uuid_str)
+
+    df.write \
+        .option("compression", compression_codec) \
+        .option("parquet.block.size", parquet_row_group_size_bytes) \
+        .parquet(save_to_dir)
+    atexit.register(_delete_cache_data_atexit, save_to_dir)
```

Comment:
It is possible that this `atexit` hook won't get executed if the interpreter is killed suddenly.  We should probably check the cache dir to see if anything is already in there that's not known to the in-memory cache.  Alternatively, you can persist the in-memory cache contents to a file and then delete everything in that file the first time this is next used.

We should also think about cases where the user might not want their cached data to be deleted.  In that case, we might want the `atexit` hook to be optional subject to a parameter passed in to `make_spark_converter`.

Reply:
CC @mengxr WDYT ?

Reply:
We should definitely make it clear that the atexit handler is a best effort. Users shouldn't rely on it for retention of sensitive data. For example, they can configure S3 lifecycle rules. Writing to the same location might not work in a multi-tenant environment, because it is hard to know whether the cached folder is still being used by some other Spark applications or not. To help manage the data, shall we at least use a date prefix?

@WeichenXu123, can we trust the semantic equality across different Spark applications?

Reply:
@mengxr
>  can we trust the semantic equality across different Spark applications?

Different Spark applications will have its own driver process (even they share the same spark cluster), and each application has its own in-memory meta-cache table (each entry is "df-plan -> df caching dir with UUID naming"), so that don't need to care about semantic equality across different Spark applications.

Reply:
Just to clarify, this means we cannot reliably tell whether two DataFrames from two different Spark applications are same, correct?

Reply:
> just to clarify, this means we cannot reliably tell whether two DataFrames from two different Spark applications are same, correct?
Yes